### ISSUE TITLE
Computer freeze when feeding a large numpy array as input in MNIST tutorial

### ISSUE THREAD

**Issue by facundoq** [tensorflow/tensorflow#6019]
### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?
None

### Environment info
intel i5, 8gb ram
Operating System:
Ubuntu 14.04.5:
Linux 4.4.0-45-generic #66~14.04.1-Ubuntu SMP Wed Oct 19 15:05:38 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

Installed version of CUDA and cuDNN: 
No cuda or cuDNN, running on CPU

If installed from binary pip package, provide:

1. A link to the pip package you installed:
TF_BINARY_URL=https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-0.11.0-cp34-cp34m-linux_x86_64.whl
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.
0.11.0rc2

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)
Use the mnist tutorial, replace
    `train_accuracy = accuracy.eval(feed_dict={
        x:batch[0], y_: batch[1], keep_prob: 1.0})`
with 
`train_accuracy = accuracy.eval(feed_dict={ x:mnist.train.images, y_: mnist.train.labels, keep_prob: 1.0})`

So that the accuracy is evaluated over the entire training set. I've reproduced this using another dataset, and the problem goes away when using a smaller number of examples.
(pastebin for entire file with this modification: http://pastebin.com/THNqB4ws)

### What other attempted solutions have you tried?
None, there's no error message and the computer hangs the first time it executes `accuracy.eval`

**Comment by poxvoculi** [comment]
The documentation (https://www.tensorflow.org/versions/r0.12/tutorials/mnist/beginners/index.html) says that mnist.test.images is 10k images, while mnist.train.images is 55k images.  By feeding the entire training set at once, it seems likely that you're causing too much memory to be allocated, and that may be your problem.  8GB is not a lot of RAM these days.   In linux you can use ulimit to bound the memory permitted to a process.

**Comment by facundoq** [comment]
@poxvoculi Thanks, I guessed as much, but just before the freeze I still see 3gbs free of ram; maybe a big chunk gets allocated just before the freeze and causes it. Also, I thought having no swap would cause an out-of-memory error, my bad, so maybe that could be it.  I'll check tomorrow from the computer at work if limiting the memory for the process helps to avoid the freeze.  

Nonetheless, the entire training set is 160mb if in float32. The output of the convolutional layers are 1.3gb and 650mb respectively. 214mb for the first fully connected and just 2 megs for the output, for a total of ~2.5gb total. The memory required for the weights is negligible. It seems weird the activations of the intermediate layers would take that much space.

**Comment by facundoq** [comment]
Indeed limiting the memory of the process avoids the OOM. TF is trying to use about 11gbs of memory to evaluate the entire training set. While this can be workaround quite easily (evaluating the accuracy in smaller batches), isn't 11gbs a bit overmuch? Is it keeping the activations of every layer even when not doing a train_step.run()? If the activations are not saved for the backward pass, no more than 1.3gb+160mb should be used by the forward pass. Is there any way to avoid storing the activations?

**Comment by facundoq** [comment]
Further testing with a really simple network: conv(4) - pool(2x2) - fc(10) (that's 4 filters for the convolutional layer) full code available in http://pastebin.com/Zh14pS3C.
I removed the training, accuracy evaluation, and most layers from the net, while also reducing the filters for the convolutional layer and the relu. For the whole training dataset this still requires **6.5gb** of memory. There's a roughly 1gb decrease if I evaluate with 10000 less training examples, 2gb for 20000, and so on. Is this normal?

**Comment by yaroslavvb** [comment]
What happens if you run with tcmalloc? There's been some memory leaks
reported earlier which disappear when you do tcmalloc (as specified here
<http://goog-perftools.sourceforge.net/doc/tcmalloc.html>)
Also, you can see which ops are allocating memory by doing something like
this

run_metadata = tf.RunMetadata()
sess.run([C.op],
             options=tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE),
             run_metadata=run_metadata)
print run_metadata

Or more visually in Chrome Trace Viewer (see
https://github.com/tensorflow/tensorflow/issues/1824#issuecomment-225754659)

trace = timeline.Timeline(step_stats=run_metadata.step_stats)
with open('timeline.ctf.json', 'w') as trace_file:
  trace_file.write(trace.generate_chrome_trace_format(show_memory=True))

**Comment by facundoq** [comment]
@yaroslavvb Running with tcmalloc decreased the memory usage slightly but not entirely.

I tried the tracer but it seems there's only cpu use info.

Here's the output from `print(run_metadata)`, http://pastebin.com/tXwB7Lfn when testing with **4000** training examples (so that the run() completes).

The relevant? bits from the memory subsection:
The Conv2D allocates `total_bytes: 25088000` (~25mb) (28*28*4000*4*2) 
The Add allocates `total_bytes: 25088000` (~25mb) (28*28*4000*4*2)
The MaxPool allocates `total_bytes: 6272000` (~6mb) (14*14*4000*4*2)
The MatMul allocates `total_bytes: 160000` (~160k) (10*4000*4)
The Add_1 allocates `total_bytes: 160000` (~160k) (10*4000*4)
The SoftmaxCrossEntropyWithLogits allocates `total_bytes: 192000` (~192k) (10*4000*4+4000*8)

There are also a lot of  "Shape", "Slice", most of which allocate just 4, 8, 16 or 200 bytes. Other nodes of the sort allocate bytes 15680, 15680, 12544000, 6272000, 160000, 160000, but all of these add up to at most 20mb. So if with 4000 samples the "use" is less than 100mb, with 55000 it would be about 1350mb or 1.35gb.

**Comment by yaroslavvb** [comment]
Uncompressed MNIST dataset is about 50MB, and you are feeding the whole dataset at each step for 20k steps. This [line](https://github.com/tensorflow/tensorflow/blob/d42facc3cc9611f0c9722c81551a7404a0bd3f6b/tensorflow/python/client/tf_session_helper.cc#L445) in session_helper.py does a copy of numpy data into TensorFlow

`std::memcpy(data, PyArray_DATA(array), size)`

So during the run of this program you are trying to copy 1TB worth of numpy arrays into TensorFlow

**Comment by facundoq** [comment]
Well, if you check the MWE you'll see that I removed the for loop and I'm doing a single eval, so I guess that theory is irrelevant for this issue.

 Regardless, you are telling me that each time I feed a numpy array to a graph memory gets allocated and never released?

**Comment by yaroslavvb** [comment]
I'm not able to reproduce freezing after single eval. When I ran your example from http://pastebin.com/THNqB4ws it runs for >100 iterations, but it was using all cores and growing memory over 30GB, so I killed it. I've seen cases of a machine freezing to a point of requiring reboot when TensorFlow allocated too much memory.

If I put a pause after 100 iterations in your script, the memory usage returns to normal, so I guess that memory gets garbage collected eventually.

**Comment by yaroslavvb** [comment]
Here's a version of your script with some extra info printed: http://pastebin.com/Sfhn3sAP
When I run it, I see this:

```
Iteration 0 time 81.24 3.31
tcmalloc: large alloc 5017829376 bytes == 0xc3be0000 @  0x7f5cb6236c4c 0x7f5cb6255040 0x7f5c50ac253f 0x7f5c517604f8 0x7f5c5176cd7d 0x7f5c5177ba14 0x7f5c5177dd23 0x7f5c51d2a49e 0x7f5c51d1dca0 0x7f5c51efb668 0x7f5c51efae42 0x7f5cb49bf870 0x7f5cb5b1e184 0x7f5cb4f3637d (nil)
test accuracy 0.1052
EIteration 0 time 1790.80 10.95
Iteration 10 time 41.54 10.95
Iteration 20 time 41.43 10.95
Iteration 30 time 40.35 10.95
Iteration 40 time 41.02 10.95
Iteration 50 time 39.61 10.95
Iteration 60 time 42.82 10.95
Iteration 70 time 41.60 10.95
Iteration 80 time 39.79 10.95
Iteration 90 time 41.15 10.95
Iteration 100 time 42.20 10.95
tcmalloc: large alloc 5017829376 bytes == 0x6825e000 @  0x7f5cb6236c4c 0x7f5cb6255040 0x7f5c50ac253f 0x7f5c517604f8 0x7f5c5176cd7d 0x7f5c5177ba14 0x7f5c5177dd23 0x7f5c51d2a49e 0x7f5c51d1dca0 0x7f5c51efb668 0x7f5c51efae42 0x7f5cb49bf870 0x7f5cb5b1e184 0x7f5cb4f3637d (nil)
test accuracy 0.7846
EIteration 100 time 1668.47 10.95
Iteration 110 time 43.09 10.95
Iteration 120 time 47.44 10.95
Iteration 130 time 49.22 10.95
Iteration 140 time 45.06 10.95
Iteration 150 time 45.78 10.95
Iteration 160 time 44.96 10.95
Iteration 170 time 45.14 10.95
Iteration 180 time 45.68 10.95
Iteration 190 time 40.74 10.95
Iteration 200 time 41.44 10.95
tcmalloc: large alloc 5017829376 bytes == 0xc3bde000 @  0x7f5cb6236c4c 0x7f5cb6255040 0x7f5c50ac253f 0x7f5c517604f8 0x7f5c5176cd7d 0x7f5c5177ba14 0x7f5c5177dd23 0x7f5c51d2a49e 0x7f5c51d1dca0 0x7f5c51efb668 0x7f5c51efae42 0x7f5cb49bf870 0x7f5cb5b1e184 0x7f5cb4f3637d (nil)
test accuracy 0.8924
EIteration 200 time 1707.17 10.95
Iteration 210 time 40.57 10.95
Iteration 220 time 44.48 10.95
Iteration 230 time 42.95 10.95
Iteration 240 time 40.86 10.95
Iteration 250 time 42.96 10.95
Iteration 260 time 42.18 10.95
Iteration 270 time 40.89 10.95
Iteration 280 time 42.90 10.95
Iteration 290 time 40.68 10.95
Iteration 300 time 41.88 10.95
tcmalloc: large alloc 5017829376 bytes == 0xc3bde000 @  0x7f5cb6236c4c 0x7f5cb6255040 0x7f5c50ac253f 0x7f5c517604f8 0x7f5c5176cd7d 0x7f5c5177ba14 0x7f5c5177dd23 0x7f5c51d2a49e 0x7f5c51d1dca0 0x7f5c51efb668 0x7f5c51efae42 0x7f5cb49bf870 0x7f5cb5b1e184 0x7f5cb4f3637d (nil)
test accuracy 0.9203
EIteration 300 time 1684.26 10.95
Iteration 310 time 51.82 10.95
Iteration 320 time 41.72 10.95
Iteration 330 time 40.68 10.95
Iteration 340 time 43.42 10.95
Iteration 350 time 48.30 10.95
Iteration 360 time 45.63 10.95
Iteration 370 time 43.82 10.95
Iteration 380 time 41.13 10.95
Iteration 390 time 41.07 10.95
Iteration 400 time 40.19 10.95
tcmalloc: large alloc 5017829376 bytes == 0xc3bde000 @  0x7f5cb6236c4c 0x7f5cb6255040 0x7f5c50ac253f 0x7f5c517604f8 0x7f5c5176cd7d 0x7f5c5177ba14 0x7f5c5177dd23 0x7f5c51d2a49e 0x7f5c51d1dca0 0x7f5c51efb668 0x7f5c51efae42 0x7f5cb49bf870 0x7f5cb5b1e184 0x7f5cb4f3637d (nil)
```

So it seems to be executing normally without slowdown.

**Comment by yaroslavvb** [comment]
I took a [closer look](https://github.com/yaroslavvb/notebooks/blob/master/mnist-memory.ipynb) at your memory usage and the main source of memory usage is the activation in your first conv layer.

Your convolution has 32 filters, so activations for first conv layer takes 4x28x28x32 = 100k per example. This is followed by broadcasting add which makes it 200k total. To run this over whole dataset you need 60k x 200k = 12GB of RAM

**Comment by facundoq** [comment]
Thanks Yarsolav, that may be it. I thought the forward pass would be smarter and discard intermediate computations or do them in-place when not needed for the backward pass, but thinking about it a bit more that may be hard to do with a framework as general as tf.

**Comment by yaroslavvb** [comment]
Actually if you look at [memory timeline](https://github.com/yaroslavvb/notebooks/blob/master/mnist-memory.ipynb) you'll see that activations for `h_conv1` are discarded as soon as `h_conv1+b_conv1` completes. This is because it's eval pass, so they are not needed for derivatives. If `add` could be done in place this would essentially lower the peak usage from 12GB to 6GB, and that's perhaps what the upcoming XLA framework could do. However, it would be trickier to lower usage for the backward pass as well since `h_conv1` activations are needed until much later -- you would need an implementation of fused `conv + add` op, and a corresponding gradient.

**Comment by facundoq** [comment]
Yes, a fused conv+add could store the activation A(b_conv1), and do A(b_conv1)-b_conv1 in the backward pass (again, in-place) to obtain A(h_conv1) again! That would surely help for big models even with small batch sizes. 

Anyhow, in this case I guess the sane thing to do is to use a tf input queue that would automatically batch stuff or calculate the full training set accuracy by feeding small batches which is easy since it's a reduction op.

**Comment by kargarisaac** [comment]
I have the same problem. I cannot load the numpy array into memory (it's not mnist) it freezes. I think I have to split my numpy array into some smaller arrays and use a generator to load every one of them and feed to my model but I think it's too slow. Is there any way to handle this problem?

**Comment by gauss-clb** [comment]
It is the best solution for OOM, the activations occupied the great majority of GPU memory, rather than images of test dataset.  But too much batch size will make activations become huge, and get the OOM.
